### PR TITLE
Store limited number of models and clear older ones

### PR DESCRIPTION
**Pull request by MukuFlash03**
Clearing or trimming number of models for a particular user from the Stage_updateable_models collection.
Constant K_MODELS_COUNT defines the number of models to keep.
Comparing with write_ts of Kth model entry to decide how many models to remove which are older than this write_ts.

Currently tested with model data available in stage dataset snapshot.
DONE - Testing done by generating mock models and cleaning them up.

### CODE REVIEW

File: emission/analysis/classification/inference/labels/inferrers.py
Comment (shankari) [inline emission/analysis/classification/inference/labels/inferrers.py:9]: why do we need this import?

File: emission/analysis/classification/inference/labels/inferrers.py
Comment (shankari) [inline emission/analysis/classification/inference/labels/inferrers.py:1]: these are additional log statements, unrelated to storing a limited number of models.
Please move into separate PR

File: emission/storage/modifiable/builtin_model_storage.py
Comment (shankari) [inline emission/storage/modifiable/builtin_model_storage.py:9]: what do we need this import for?

**Review by shankari — CHANGES_REQUESTED** [review]
I am stopping at this point because there are multiple comments related to , please clean up this PR and re-submit for review.

File: .docker/setup_config.sh
Comment (MukuFlash03) [reply .docker/setup_config.sh:1]: Removed these changes.

This PR was opened prior to the latest merged changes for the security fixes.
And then I continued working on them after the merged security fixes.
Hence, I thought of keeping the code updated and pulled the latest changes from master.
But yes, you would have been able to see those changes anyways when you would have merged the code.

File: emission/analysis/classification/inference/labels/inferrers.py
Comment (MukuFlash03) [reply emission/analysis/classification/inference/labels/inferrers.py]: Might have added to test locally, removed from latest fixes.

File: emission/storage/modifiable/builtin_model_storage.py
Comment (MukuFlash03) [reply emission/storage/modifiable/builtin_model_storage.py]: Might have added to test locally, but is currently an unused import.
Removed from latest fixes.

File: emission/storage/modifiable/builtin_model_storage.py
Comment (shankari) [inline emission/storage/modifiable/builtin_model_storage.py:26]: why is this still a TODO? You have just done this, right?

File: emission/storage/modifiable/builtin_model_storage.py
Comment (shankari) [inline emission/storage/modifiable/builtin_model_storage.py:32]: I think it would be more useful to log in the change the number of modules - e.g. something like

```suggestion
        logging.debug("model count for user %s has changed %s -> %s" % (self.user_id, old_model_count, new_model_count))
```

File: emission/storage/modifiable/builtin_model_storage.py
Comment (shankari) [inline emission/storage/modifiable/builtin_model_storage.py:85]: Or maybe this is where we need the `old -> new` log. In that case, I am not sure what the previous log adds.

File: emission/tests/storageTests/TestModelStorage.py
Comment (shankari) [inline emission/tests/storageTests/TestModelStorage.py:53]: this should not be true if we tear down properly. what was the set of steps that led to the data being saved between test invocations?

File: emission/tests/storageTests/TestModelStorage.py
Comment (shankari) [inline emission/tests/storageTests/TestModelStorage.py:87]: when will you run into this situation? If this can happen frequently and we fail instead of retrying when it happens, we may have a non-deterministic test, which is not a great idea for an automated test pipeline.

File: emission/tests/storageTests/TestModelStorage.py
Comment (shankari) [inline emission/tests/storageTests/TestModelStorage.py:120]: why is this hardcoded to 20?

**Review by shankari — CHANGES_REQUESTED** [review]
Looks good at a high level; but I would like some polishing changes...

File: emission/storage/modifiable/builtin_model_storage.py
Comment (MukuFlash03) [reply emission/storage/modifiable/builtin_model_storage.py:85]: Yes, it makes more sense to add the change in model count here.
I'll explain the need for the logging statement present at the beginning in `upsert_model()`.
To clarify the three different states where the logging statements are present, say we have 3 models currently.

Two sets of changes to model count are there: a) trimming models b) inserting new model

1. Current count before **trimming** models = 3
```
# In trim_model_entries()
current_model_count = edb.get_model_db().count_documents({"user_id": self.user_id})
logging.debug("Before trimming, model count for user %s = %s" % (self.user_id, current_model_count))
```

2. New count after **trimming** models = 2
```
# In trim_model_entries()
new_model_count = edb.get_model_db().count_documents({"user_id": self.user_id})
logging.debug("After trimming, model count for user %s = %s" % (self.user_id, new_model_count))
```

3. New count after **inserting** models = 3
```
# In upsert_model()
new_model_count = edb.get_model_db().count_documents({"user_id": self.user_id})
logging.debug("New model count for user %s = %s" % (self.user_id, new_model_count))
```

File: emission/tests/storageTests/TestModelStorage.py
Comment (MukuFlash03) [reply emission/tests/storageTests/TestModelStorage.py:53]: So, I've taken this code from an existing test file - [TestRunGreedyModel.py](https://github.com/e-mission/e-mission-server/blob/master/emission/tests/modellingTests/TestRunGreedyModel.py), as I wasn't really sure how to generate models, trips and didn't want to re-write the code and miss any important parameters, variables needed for training, inserting models.

But I'll make changes as per my understanding.

File: emission/tests/storageTests/TestModelStorage.py
Comment (MukuFlash03) [reply emission/tests/storageTests/TestModelStorage.py:87]: Answered [here](https://github.com/e-mission/e-mission-server/pull/948#discussion_r1478739529).
But I'll make changes as per my understanding.

File: emission/tests/storageTests/TestModelStorage.py
Comment (shankari) [reply emission/tests/storageTests/TestModelStorage.py:113]: Right, but I would have expected to see something like `maximum_stored_model_count + 2` to be more flexible. I will accept this for now since this is part of the unit tests and the sample value is not overwritten here.

File: emission/tests/storageTests/TestModelStorage.py
Comment (shankari) [reply emission/tests/storageTests/TestModelStorage.py]: It is fine to keep it. We likely have not run into this error before, but it is not bad to have that check, as long as it is well documented.

File: emission/analysis/modelling/trip_model/config.py
Comment (shankari) [inline emission/analysis/modelling/trip_model/config.py:78]: I'm not going to hold up the merge for this release, but would be good to avoid these extraneous whitespace changes in future PRs

File: emission/storage/modifiable/builtin_model_storage.py
Comment (shankari) [inline emission/storage/modifiable/builtin_model_storage.py:9]: like this

**Review by shankari — CHANGES_REQUESTED** [review]
Where in here are you checking that the models that are deleted are the oldest ones? I see that the code does that with the sort on the `metadata.write_ts` but I don't see a test for that. If I changed 

```
edb.get_model_db().find(find_query).sort("metadata.write_ts", -1)
```

to 

```
edb.get_model_db().find(find_query).sort("metadata.write_ts", 1)
```

I believe the test would still pass

File: emission/tests/storageTests/TestModelStorage.py
Comment (MukuFlash03) [reply emission/tests/storageTests/TestModelStorage.py]: Oh, understood, hardcoding was probably not the best idea. Changed this.

File: emission/analysis/modelling/trip_model/config.py
Comment (MukuFlash03) [reply emission/analysis/modelling/trip_model/config.py:78]: Oh, right. It's just that sometimes there are extra lines at the end of the file and when I'm adding new code I tend to keep the spacing between modules similar to how it is in other parts of the file. But will try to minimize these changes.

**Comment by MukuFlash03** [comment]
> Where in here are you checking that the models that are deleted are the oldest ones? I see that the code does that with the sort on the `metadata.write_ts` but I don't see a test for that. If I changed
> 
> ```
> edb.get_model_db().find(find_query).sort("metadata.write_ts", -1)
> ```
> 
> to
> 
> ```
> edb.get_model_db().find(find_query).sort("metadata.write_ts", 1)
> ```
> 
> I believe the test would still pass

Thank you for pointing that out; that is indeed an important test I missed.

For testing purposes, I did change the sort(-1) to sort(1) which leads to ascending order of queried models based on write_ts.
I had the following observations :
- Initial count assertion itself fails (which is good!) as all available models satisfy `write_ts` condition of `$ lte < write_ts_limit` .
```
            filter_clause = {
                "user_id" : self.user_id,
                "metadata.key" : key,
                "metadata.write_ts" : { "$lte" : write_ts_limit }
            }
```

- Hence every time all models are being deleted and only one model remains (current_model_count = 1) which is the latest one being inserted.
```
if i <= (maximum_stored_model_count - 1):
    self.assertEqual(current_model_count, i+1)
else:
    self.assertEqual(current_model_count, maximum_stored_model_count)
```

- Just to get beyond this point, I tried testing without this assertion, to observe `write_ts_lists`:
    - For every `(maximum_stored_model_count + 1)th` model added, the first `maximum_stored_model_count ` models are deleted and the latest model is inserted and this repeats.

--------- 

So, to reinforce the testing, I’m adding another assertion to confirm that only the latest top K or max_model_count models are stored by storing the write_ts times in two different lists and checking for equality of the top K entries:
```
- model_creation_write_ts_list : stores write_ts times each time a model is created in the for loop.
- stored_model_write_ts_list : stores write_ts times of all the already stored models in the DB, which should just have the latest models.
- The last 'maximum_stored_model_count' in model_creation_write_ts_list should match those in stored_model_write_ts_list.
```

**Comment by shankari** [comment]
I do not 100% understand how the first assertion tripped, but I am satisfied with the changes and am happy to merge!

**Review by shankari — APPROVED** [review]

**Pull request merged**
